File: java/src/main/java/problems101_200/MinStack.java
package problems101_200;

import java.util.Stack;

class MinStack {
    Stack<Integer> main = new Stack<>();
    Stack<Integer> mins = new Stack<>();

    void push(int value) {
        main.push(value);
        if (value <= getMinimum()) {
            mins.push(value);
        }
    }

    int getMinimum() {
        return mins.isEmpty()
            ? Integer.MAX_VALUE
            : mins.peek();
    }

    void pop() {
        int value = main.pop();
        if (value == getMinimum()) {
            mins.pop();
        }
    }

    int top() {
        return main.peek();
    }
}

File: kotlin/src/main/kotlin/problems101_200/MinStack.kt
package problems101_200

import java.util.Stack

class MinStack {
    var main = Stack<Int>()
    var mins = Stack<Int>()

    fun push(value: Int) {
        main.push(value)
        if (value <= minimum) {
            mins.push(value)
        }
    }

    val minimum: Int
        get() =
            if (mins.isEmpty()) {
                Int.MAX_VALUE
            } else {
                mins.peek()
            }

    fun pop() {
        val value = main.pop()
        if (value == minimum) {
            mins.pop()
        }
    }

    fun top(): Int = main.peek()
}
